File: medium/22_generate_parenthesis.rb
# 22. Generate Parentheses
# Medium
#
# Given n pairs of parentheses, write a function to generate all combinations
# of well-formed parentheses.

def generate_parenthesis(n)
  helper(n, n, [])
end

def helper(m, n, s)
  return [s.join] if m.zero? && n.zero?

  l1 = m.positive? ? helper(m - 1, n, s.clone << '(') : []
  l2 = n > m ? helper(m, n - 1, s << ')') : []

  l1.concat(l2)
end

p generate_parenthesis(3) # ["((()))","(()())","(())()","()(())","()()()"]
p generate_parenthesis(1) # ["()"]

File: python/medium/22_generate_parenthesis.py
"""
22. Generate Parentheses
Medium

Given n pairs of parentheses, write a function to generate all combinations
of well-formed parentheses.
"""


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        return helper(n, n, [])


def helper(m, n, ps):
    if m == 0 and n == 0:
        return ["".join(ps)]

    ps1, ps2 = [], []
    if m > 0:
        ps1 = helper(m - 1, n, ps + ["("])
    if n > m:
        ps2 = helper(m, n - 1, ps + [")"])

    return ps1 + ps2
